GH-38837: [Format] Add the specification to pass statistics through the Arrow C data interface (#43553)
apache/arrow: docs/source/format/CDataInterfaceStatistics.rst

.. or more contributor license agreements.  See the NOTICE file
.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

.. _c-data-interface-statistics:

=====================================================
Passing statistics through the Arrow C data interface
=====================================================

Rationale
=========

Statistics are useful for fast query processing. Many query engines
use statistics to optimize their query plan.

Apache Arrow format doesn't have statistics but other formats that can

> be read as Apache Arrow data may have statistics. For example, Apache
> Parquet C++ can read Apache Parquet file as Apache Arrow data and
> Apache Parquet file may have statistics.

Reviewer: ```suggestion
be read as Apache Arrow data may have statistics. For example, the
Apache Parquet C++ implementation can read an Apache Parquet file as
Apache Arrow data and the Apache Parquet file may have statistics.
```

One of the Arrow C data interface use cases is the following:

> 1. Module A reads Apache Parquet file as Apache Arrow data
> 2. Module A passes the read Apache Arrow data to module B through the
>    Arrow C data interface
> 3. Module B processes the passed Apache Arrow data

Reviewer: ```suggestion
1. Module A reads Apache Parquet file as Apache Arrow data.
2. Module A passes the read Apache Arrow data to module B through the
   Arrow C data interface.
3. Module B processes the passed Apache Arrow data.
```

If module A can pass the statistics associated with the Apache Parquet
file to module B through the Arrow C data interface, module B can use
the statistics to optimize its query plan.

Goals
-----

> * Provide the standard way to pass statistics through the Arrow C data
>   interface to avoid reinventing the wheel.
> * The standard way must be easy to use with the Arrow C data interface.

Reviewer: ```suggestion
* Establish a standard way to pass statistics through the Arrow C data
  interface.
* Provide this in a manner that enables compatibility and ease of
  implementation for users of the Arrow C data interface.
```

Non-goals
---------

* Provide a common way to pass statistics that can be used for

>   other interfaces such Arrow Flight too.

Reviewer: What about the Arrow IPC format? Can you add a sentence here that explains why we do not recommend using this to pass statistics over Arrow IPC?

Author: Good point. This may fit the Arrow IPC format. (A producer sends data and statistics as 2 separated the Arrow IPC format data.)

But the Arrow IPC format can use more approaches. For example, the Arrow IPC format can have metadata for each record batch data: https://arrow.apache.org/docs/format/Columnar.html#encapsulated-message-format (The Arrow C data can't have metadata for `ArrowArray`.) 
The Arrow IPC format can be used with other mechanisms such as Arrow Flight and ADBC.

So this may not be the best approach for the Arrow IPC format. We should discuss this use case with the Arrow IPC format separately.

I'll add something to here.

For example, ADBC has the statistics related APIs. This specification

> doesn't replace them.

Reviewer: Maybe link to https://arrow.apache.org/adbc/current/format/specification.html#statistics

Author: Thanks. I should have done it.

.. _c-data-interface-statistics-schema:

Schema
======

> This specification provides only the schema for statistics. Producers
> passes statistics as a map Arrow array that uses the schema through
> the Arrow C data interface.

Reviewer: ```suggestion
This specification provides only the schema for statistics. The
producer passes statistics through the Arrow C data interface as an
Arrow map array that uses this schema.
```

> Here is the schema for a statistics map Arrow Array:

Reviewer: ```suggestion
Here is the schema for a statistics Arrow map array:
```

.. list-table::
   :header-rows: 1

   * - Key or items
     - Data type
     - Nullable
     - Notes
   * - key
     - ``int32``
     - ``true``

>      - The column index or null if the statistics refer to whole table
>        or record batch.

Reviewer: ```suggestion
     - The column index, or null if the statistics refer to whole
       table or record batch.
```

Reviewer: The index is zero-based, correct?

Author: Correct.

   * - items
     - ``map``
     - ``false``
     - Statistics for the target column, table or record batch. See

>        the separated table for details.

Reviewer: ```suggestion
       the separate table below for details.
```

Here is the schema for the statistics map:

.. list-table::
   :header-rows: 1

   * - Key or items
     - Data type
     - Nullable
     - Notes
   * - key
     - ``dictionary<int32, utf8>``
     - ``false``
     - Statistics key is string. Dictionary is used for
       efficiency. Different keys are assigned for exact value and

>        approximate value. See also the separated description for
>        statistics key.

Reviewer: ```suggestion
       approximate value. Also see the separate description below for
       statistics key.
```

   * - items
     - ``dense_union``
     - ``false``
     - Statistics value is dense union. It has at least all needed
       types based on statistics kinds in the keys. For example, you
       need at least ``int64`` and ``float64`` types when you have a
       ``int64`` distinct count statistic and a ``float64`` average

>        byte width statistic. See also the separated description for
>        statistics key.

Reviewer: ```suggestion
       byte width statistic. Also see the separate description below
       for statistics key.
```

       We don't standardize field names for the dense union because
       consumers can access to proper field by index not name. So
       producers can use any valid name for fields.

.. _c-data-interface-statistics-key:

Statistics key
--------------

Statistics key is string. ``dictionary<int32, utf8>`` is used for
efficiency.

We assign different statistics keys for variants instead of using
flags. For example, we assign different statistics keys for exact
value and approximate value.

The colon symbol ``:`` is to be used as a namespace separator like
:ref:`format_metadata`. It can be used multiple times in a key.

The ``ARROW`` pattern is a reserved namespace for pre-defined
statistics keys. User-defined statistics must not use it.

> Here are pre-defined statistics keys:

Author: They are the same as ADBC's one: https://github.com/apache/arrow-adbc/blob/05fa60d643c66b572d426ab28aa78fc52e9520e8/c/include/arrow-adbc/adbc.h#L524-L570

  other interfaces such Arrow Flight too.

For example, ADBC has `the statistics related APIs
<https://arrow.apache.org/adbc/current/format/specification.html#statistics>`__.
This specification doesn't replace them.

This specification may fit some use cases of :ref:`format-ipc` not the
Arrow data interface. But we don't recommend this specification for
the Arrow IPC format for now. Because we may be able to define better
specification for the Arrow IPC format. The Arrow IPC format has some
different features compared with the Arrow C data interface. For
example, the Arrow IPC format can have :ref:`ipc-message-format
metadata for each message`. If you're interested in the specification
for passing statistics through the Arrow IPC format, please start a
discussion on the `Arrow development mailing-list
<https://arrow.apache.org/community/>`__.

.. _c-data-interface-statistics-schema:

Schema
======

This specification provides only the schema for statistics. The
producer passes statistics through the Arrow C data interface as an
Arrow map array that uses this schema.

Here is the outline of the schema for statistics::

    struct<
      column: int32,

>       values: map<

Author: `values` isn't a good name because it's a map array of statistics key and statistics value pairs.
`statistics` (or `items`) may be better.

Author: `values` -> `statistics`.

   * - ``ARROW:distinct_count:exact``
     - ``int64``
     - The number of distinct values in the target. (exact)
   * - ``ARROW:distinct_count:approximate``
     - ``float64``
     - The number of distinct values in the target. (approximate)
   * - ``ARROW:max_byte_width:exact``
     - ``int64``
     - The maximum size in bytes of a row in the target. (exact)
   * - ``ARROW:max_byte_width:approximate``
     - ``float64``
     - The maximum size in bytes of a row in the target. (approximate)
   * - ``ARROW:max_value:exact``
     - Target dependent
     - The maximum value in the target. (exact)
   * - ``ARROW:max_value:approximate``
     - Target dependent
     - The maximum value in the target. (approximate)
   * - ``ARROW:min_value:exact``
     - Target dependent
     - The minimum value in the target. (exact)
   * - ``ARROW:min_value:approximate``
     - Target dependent
     - The minimum value in the target. (approximate)
   * - ``ARROW:row_count:exact``
     - ``int64``
     - The number of rows in the target table or record batch. (exact)
   * - ``ARROW:row_count:approximate``
     - ``float64``
     - The number of rows in the target table or record

>        batch. (approximate)

Author: `ARROW:null_count:{exact,approximate}` are missing.
I'll add them later.

Author: Added.

This specification doesn't replace them.

This specification may fit some use cases of :ref:`format-ipc` not the
Arrow data interface. But we don't recommend this specification for
the Arrow IPC format for now. Because we may be able to define better
specification for the Arrow IPC format. The Arrow IPC format has some
different features compared with the Arrow C data interface. For
example, the Arrow IPC format can have :ref:`ipc-message-format
metadata for each message`. If you're interested in the specification
for passing statistics through the Arrow IPC format, please start a
discussion on the `Arrow development mailing-list
<https://arrow.apache.org/community/>`__.

.. _c-data-interface-statistics-schema:

Schema
======

This specification provides only the schema for statistics. The
producer passes statistics through the Arrow C data interface as an
Arrow map array that uses this schema.

Here is the outline of the schema for statistics::

    map<
      key: int32,
      items: map<
        key: dictionary<
          indices: int32,
          dictionary: utf8
        >,
        items: dense_union<...all needed types...>,
      >

>     >

Reviewer 2: Does this mean we get an array of maps or a single map? Because we really only need a single map. Any consumer of these statistics will loop through the values and gather only the statistic values that it cares about by skipping the keys that it doesn't understand.

Author: An (one element) array of maps. But we can simplify this. I didn't notice it. Thanks.

How about the following? It removes the outer `map`:

    column: int32,
    values: map<
      key: dictionary<
        indices: int32,
        dictionary: utf8
      >,
      items: dense_union<...all needed types...>,
    >

Author: I've updated this part!

.. Licensed to the Apache Software Foundation (ASF) under one
.. or more contributor license agreements.  See the NOTICE file
.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

.. _c-data-interface-statistics:

=====================================================

> Passing statistics through the Arrow C data interface

Reviewer 3: Any Arrow array can be transmitted over Flight or IPC, including statistics arrays. I agree with @lidavidm about explaining the context and original motivation, but I don't think we should exclude IPC or Flight here. People can use statistics arrays over these protocols if they find it useful.

> Passing statistics through the Arrow C data interface

Reviewer 4: I think the confusion is that I think Kou is trying to say that this isn't the "one true way" to embed statistics inside Flight or an IPC file (or in other words, this isn't a spec for adding Parquet-like statistics to IPC), but of course there's nothing stopping you from using those APIs to transmit statistics data for some other purpose. (Correct me if I'm wrong here @kou.) Again, because the original motivation was the DuckDB use case and so trying to solve "statistics for IPC files" was out of scope.

> Passing statistics through the Arrow C data interface

Author: Thanks! It's correct.

(I'll be able to work on this proposal tomorrow. I'll update the current proposal and start a discussion on the mailing list tomorrow.)

> Passing statistics through the Arrow C data interface

Author: (Sorry... I couldn't work on this today... I hope that I can work on this tomorrow...)

be read as Apache Arrow data may have statistics. For example, the
Apache Parquet C++ implementation can read an Apache Parquet file as
Apache Arrow data and the Apache Parquet file may have statistics.

One of :ref:`c-data-interface` use cases is the following:

1. Module A reads Apache Parquet file as Apache Arrow data.
2. Module A passes the read Apache Arrow data to module B through the
   Arrow C data interface.
3. Module B processes the passed Apache Arrow data.

If module A can pass the statistics associated with the Apache Parquet
file to module B through the Arrow C data interface, module B can use
the statistics to optimize its query plan.

Goals
-----

* Establish a standard way to pass statistics through the Arrow C data
  interface.
* Provide this in a manner that enables compatibility and ease of
  implementation for existing users of the Arrow C data interface.

Non-goals
---------

* Provide a common way to pass statistics that can be used for
  other interfaces such Arrow Flight too.

For example, ADBC has `the statistics related APIs
<https://arrow.apache.org/adbc/current/format/specification.html#statistics>`__.

> This specification doesn't replace them.

Reviewer 4: In retrospect since I think nothing in ADBC actually implements the statistics properly we may choose to deprecate the existing API and down the line reintroduce it using the new schema.

This specification may fit some use cases of :ref:`format-ipc` not the
Arrow data interface. But we don't recommend this specification for
the Arrow IPC format for now. Because we may be able to define better
specification for the Arrow IPC format. The Arrow IPC format has some
different features compared with the Arrow C data interface. For
example, the Arrow IPC format can have :ref:`metadata for each message
<ipc-message-format>`. If you're interested in the specification for
passing statistics through the Arrow IPC format, please start a
discussion on the `Arrow development mailing-list
<https://arrow.apache.org/community/>`__.

.. _c-data-interface-statistics-schema:

Schema
======

This specification provides only the schema for statistics. The
producer passes statistics through the Arrow C data interface as an
Arrow map array that uses this schema.

Here is the outline of the schema for statistics::

    struct<

>       column: int32,

Reviewer 4: Maybe we should combine it with the comment about defining "column index" above...we could extend the numbering to cover nested columns (possibly down the line)?

>       column: int32,

Author: Let's use it!

      statistics: map<
        key: dictionary<
          indices: int32,
          dictionary: utf8
        >,
        items: dense_union<...all needed types...>,
      >
    >

Here is the details of top-level ``struct``:

.. list-table::
   :header-rows: 1

   * - Name
     - Data type
     - Nullable
     - Notes
   * - ``column``
     - ``int32``
     - ``true``
     - The zero-based column index, or null if the statistics

>        describe the whole table or record batch.

Reviewer 3: We should perhaps explain what the "column index" means when the data is nested.

   * - ``statistics``
     - ``map``
     - ``false``
     - Statistics for the target column, table or record batch. See
       the separate table below for details.

Here is the details of the ``map`` of the ``statistics``:

.. list-table::
   :header-rows: 1

   * - Key or items
     - Data type
     - Nullable
     - Notes
   * - key
     - ``dictionary<indices: int32, dictionary: utf8>``
     - ``false``
     - Statistics key is string. Dictionary is used for
       efficiency. Different keys are assigned for exact value and
       approximate value. Also see the separate description below for
       statistics key.
   * - items
     - ``dense_union``
     - ``false``
     - Statistics value is dense union. It has at least all needed
       types based on statistics kinds in the keys. For example, you
       need at least ``int64`` and ``float64`` types when you have a
       ``int64`` distinct count statistic and a ``float64`` average
       byte width statistic. Also see the separate description below
       for statistics key.

       We don't standardize field names for the dense union because

>        consumers can access to proper field by index not name. So

Reviewer 3: "By index"? Do you mean by data type instead?

>        producers can use any valid name for fields.

Reviewer 3: Do we really want this? What's the point of allowing arbitrary field names? This will make interoperability more difficult.

Reviewer 4: I think the point is that consumers shouldn't rely on particular field names, given the index/type is supposed to fully define it already?

Reviewer 5: We have had trouble in the arrow-rs implementation or other places where schema's contain names but they aren't standardized

For example, we have an embedded field name for `List`s and some implementations use `"item"` and some `"element"` which causes spurious schema mismatch errors

Therefore I also recommend removing field names unless there is some good reason for doing so (it isn't clear to me from the text why there are arbitrary field namds)

Author: Thanks. I didn't notice that the schema comparison case.

In this case, I think that we don't need to care about the case. Because we have a dynamic part in the proposed statistics schema. (This dense union part.) In general, two statistics schema will not be same with filed names standardization. For example, one statistics schema use `dense_union<int64, float64>` but another schema use `dense_union<int64>`.

.. _c-data-interface-statistics-key:

Statistics key
--------------

Statistics key is string. ``dictionary<int32, utf8>`` is used for
efficiency.

> We assign different statistics keys for variants instead of using

Reviewer 3: What does "variants" mean here? Individual statistics?

> flags. For example, we assign different statistics keys for exact

Reviewer 3: What does "instead of using values" means here?

value and approximate value.

The colon symbol ``:`` is to be used as a namespace separator like
:ref:`format_metadata`. It can be used multiple times in a key.

The ``ARROW`` pattern is a reserved namespace for pre-defined

> statistics keys. User-defined statistics must not use it.

Reviewer 3: So there can be user-defined statistics? This document should be clear about it.

Here are pre-defined statistics keys:

.. list-table::
   :header-rows: 1

   * - Key
     - Data type
     - Notes
   * - ``ARROW:average_byte_width:exact``

>      - ``float``

Reviewer 3: Is "float" different from "float64"?

>      - The average size in bytes of a row in the target. (exact)

Reviewer 3: "the target" is the array, right?

Author: Done.

   * - ``ARROW:average_byte_width:approximate``
     - ``float64``
     - The average size in bytes of a row in the target. (approximate)
   * - ``ARROW:distinct_count:exact``
     - ``int64``
     - The number of distinct values in the target. (exact)
   * - ``ARROW:distinct_count:approximate``
     - ``float64``
     - The number of distinct values in the target. (approximate)
   * - ``ARROW:max_byte_width:exact``
     - ``int64``
     - The maximum size in bytes of a row in the target. (exact)
   * - ``ARROW:max_byte_width:approximate``

>      - ``float64``

Reviewer 3: Why is this float64? It doesn't seem to make sense. The maximum is obviously an integer.

Author: It's correct for the "exact" case. But I think that float64 is better for the "approximate" case. Because an approximate max byte width may be float for string array.

Reviewer 3: It doesn't make sense for the maximum approximation to be float64. The fractional part is entirely nonsensical.

     - The maximum size in bytes of a row in the target. (approximate)
   * - ``ARROW:max_value:exact``
     - Target dependent
     - The maximum value in the target. (exact)
   * - ``ARROW:max_value:approximate``
     - Target dependent
     - The maximum value in the target. (approximate)
   * - ``ARROW:min_value:exact``
     - Target dependent
     - The minimum value in the target. (exact)
   * - ``ARROW:min_value:approximate``
     - Target dependent
     - The minimum value in the target. (approximate)

>    * - ``ARROW:null_count:exact``

Reviewer 3: The C Data Interface already exposes the `null_count`, what use is this for?

Author: It's the number of `null` for the array. In this case, the array is a statistics array not a target column/record batch.
This is for a target column/record batch not a statistics array.

     - ``int64``
     - The number of nulls in the target. (exact)
   * - ``ARROW:null_count:approximate``
     - ``float64``
     - The number of nulls in the target. (approximate)
   * - ``ARROW:row_count:exact``
     - ``int64``

>      - The number of rows in the target table or record batch. (exact)

Reviewer 3: What's the point? The length is already known.

Author: It's not known yet.
In general, a statistics array is passed before we receive a target record batch.
We don't know the number of rows in a target record batch at the time.

   * - ``ARROW:row_count:approximate``
     - ``float64``
     - The number of rows in the target table or record
       batch. (approximate)

If you find a missing statistics key that is usable for multiple
systems, please propose it on the `Arrow development mailing-list
<https://arrow.apache.org/community/>`__.

Examples
========

> Here are some examples to help you understand.

Reviewer 6: Should we provide concrete examples of data in the statistics array? I'm not sure if this is the best place. An impatient reader may be interested to see what content to expect from a quick glance.

> C++

Reviewer 3: This should not be part of the general spec, can it be removed?

Author: OK. I'll remove it.

I'll add an example data record batch and statistics array for it instead.

Author: I've added one simple record batch case.
I'll add more examples later.

producer passes statistics through the Arrow C data interface as an
Arrow map array that uses this schema.

Here is the outline of the schema for statistics::

    struct<
      column: int32,
      statistics: map<
        key: dictionary<
          indices: int32,
          dictionary: utf8
        >,
        items: dense_union<...all needed types...>
      >
    >

Here is the details of top-level ``struct``:

.. list-table::
   :header-rows: 1

   * - Name
     - Data type
     - Nullable
     - Notes
   * - ``column``
     - ``int32``
     - ``true``
     - The zero-based column index, or null if the statistics
       describe the whole table or record batch.

       The column index is computed as the same rule used by
       :ref:`ipc-recordbatch-message`.
   * - ``statistics``
     - ``map``
     - ``false``
     - Statistics for the target column, table or record batch. See
       the separate table below for details.

Here is the details of the ``map`` of the ``statistics``:

.. list-table::
   :header-rows: 1

   * - Key or items
     - Data type
     - Nullable
     - Notes
   * - key
     - ``dictionary<indices: int32, dictionary: utf8>``
     - ``false``

>      - Statistics key is string. Dictionary is used for

Reviewer 5: Maybe we could be more explicitly about what the key means (it seems like it is the "statistics type" )?

Author: Thanks!
This is also pointed out by @felipecrv :
* https://github.com/apache/arrow/pull/45058#pullrequestreview-2520677048
* https://github.com/apache/arrow/pull/45058#discussion_r1895832598

https://github.com/apache/arrow/pull/45058 uses "statistics name" instead of "statistics key".

       efficiency. Different keys are assigned for exact value and
       approximate value. Also see the separate description below for
       statistics key.
   * - items
     - ``dense_union``
     - ``false``
     - Statistics value is dense union. It has at least all needed
       types based on statistics kinds in the keys. For example, you
       need at least ``int64`` and ``float64`` types when you have a
       ``int64`` distinct count statistic and a ``float64`` average
       byte width statistic. Also see the separate description below
       for statistics key.

       We don't standardize field names for the dense union because
       consumers can access to proper field by type code not name. So
       producers can use any valid name for fields.

       TODO: Should we standardize field names?

.. _c-data-interface-statistics-key:

> Statistics key

Reviewer 5: Nit is I would find the term "statistics type" rather than "statistics key" easier to understand

   * - ``ARROW:min_value:exact``
     - Target dependent
     - The minimum value in the target column. (exact)
   * - ``ARROW:min_value:approximate``
     - Target dependent
     - The minimum value in the target column. (approximate)
   * - ``ARROW:null_count:exact``
     - ``int64``
     - The number of nulls in the target column. (exact)
   * - ``ARROW:null_count:approximate``
     - ``float64``
     - The number of nulls in the target column. (approximate)
   * - ``ARROW:row_count:exact``
     - ``int64``
     - The number of rows in the target table or record batch. (exact)
   * - ``ARROW:row_count:approximate``
     - ``float64``
     - The number of rows in the target table or record
       batch. (approximate)

If you find a missing statistics key that is usable for multiple
systems, please propose it on the `Arrow development mailing-list
<https://arrow.apache.org/community/>`__.

Examples
========

Here are some examples to help you understand.

Simple record batch
-------------------

Schema::

    vendor_id: int32
    passenger_count: int64

Data::

    vendor_id:       [5, 1, 5, 1, 5]
    passenger_count: [1, 1, 2, 0, null]

> Statistics schema::

Reviewer 5: It is probably too late, but I figured I would point out another format we use for statistics in DataFusion is a columnar form rather than a nested structure 

https://docs.rs/datafusion/latest/datafusion/physical_optimizer/pruning/trait.PruningStatistics.html

For example to represent this data in DataFusion it would be

| `vendor_id::null_count` | `vendor_id::min` | `vendor_id::max` | ... | `passenger_count::max` |
|--------|--------|--------|--------|--------|
| 0 | 1 | 5 | ... | 2 |

The benefit of this encoding is that it can be used to quickly evaluate predicates on ranges (e.g. figure out if `vendor_id = 6` could ever be true)

We use this format to read statistics from the parquet files (see [`ParquetStatisticsConverter`](https://docs.rs/parquet/latest/parquet/arrow/arrow_reader/statistics/struct.StatisticsConverter.html) to rule out row groups)

It does result in potentially very wide schemas however

Author: Thanks for sharing the existing statistics representation.
This is also shared by @tustvold in the initial discussion: https://lists.apache.org/thread/hd79kp59796sqmwo24vmsnbk8t6x7h8g

We chose the current approach than the DataFusion approach because the current approach will be easy to add support for not pre-defined (application-specific) statistics names (statistics types): https://lists.apache.org/thread/g362w785ntcdp0hjjqr06lk5gl4bfrj0

    struct<
      column: int32,
      statistics: map<
        key: dictionary<
          indices: int32,
          dictionary: utf8
        >,
        items: dense_union<int64>
      >
    >

> Statistics array::

Reviewer 5: I didn't understand this example. I thought the statistics were structs, so I would have expected the data to look something like this (perhaps we could give the "logical contents" and then the specific array encoding):

```
[ 
  // first struct element
  { 
    column: null, # record batch
     statistics: {
        "ARROW:row_count:exact": 0
     }
   },
  { 
    column: 0, # vendor_id
     statistics: {
        "ARROW:null_count:exact": 0,
        "ARROW:distinct_count:exact": 2,
        "ARROW:max_value:exact": 5,
        "ARROW:min_value:exact": 1,
     }
   },
...
]
```

I can help work out the example if people think this is a good idea

Author: Let's continue this in the current PR: https://github.com/apache/arrow/pull/45058/files#r1896967442
